blog-posts/scoped-registry.md

```diff
@@ -0,0 +1,128 @@
+
+```
+@workingwombats:registry=https://registry.working-wombats.rocks/
+```
+## 401 Unauthorized... what? 😱
+
+Jeff quickly added the configuration for the scoped registries and thought he could knock off work early today and enjoy the rest of the day but then he saw the dreaded 401 error message...
+
+Even if the registry is hosted internally, it is good practice to add authentication to prevent anonymous access to the packages. This can be in form of a username and password or via an access token. For username and password normally basic authentication ([Basic Authentication](https://en.wikipedia.org/wiki/Basic_access_authentication)) is used. To create the authentication string Jeff takes the username and password, separated with a colon and then encode it with base64:
+
+`username:password` ⇒ base64 encoded ⇒ `dXNlcm5hbWU6cGFzc3dvcmQ=`
+
+For encoding you can use the command line `echo -n 'my-string' | base64` (Linux) or an IDE (vscode with plugin) or search for an online base64 encoder.
+
+>💡 Use online encoder only for testing purposes. Never enter a productive username and/or password there.
+
+Jeff remembered, that the mail of his boss also contained the username and password in plaintext to access the registry (yes, even in the best imaginary companies security isn't always perfect 🤷‍♂️):
+
+`sirWombat:theCudd1er!` ⇒ base64 encoded ⇒ `c2lyV29tYmF0OnRoZUN1ZGQxZXIh`
+
+So to add basic authentication to the registry, Jeff encodes the username and password with base64 and goes back to the .npmrc file, and adds the following line below the corresponding scoped registry:
+
+```
+@workingwombats:registry=https://registry.working-wombats.rocks/
+//registry.working-wombats.org/:_auth=c2lyV29tYmF0OnRoZUN1ZGQxZXIh
+```
+
+>💡 Note the // without the HTTPS: at the beginning
+
+Alternatively, if you have an access token you can put the access token after `_authToken`:
+
+```
+@workingwombats:registry=https://registry.working-wombats.rocks/
+//registry.working-wombats.org/:_authToken=c2lyd29tYmF0OnRoZWN1ZGQxZXIh
+```
+
+Jeff tried it again et voilà it worked! He almost committed the new additions of the project-level .npmrc to git but then his old mentor "Sir Wombat" came into his mind:
+
+>💡 Remember that it is not a good practice to add credentials to version control e.g. if you have versioned your .npmrc file.
+
+and Jeff was asking himself...
+## But where should I put the credentials then? 🤷‍♂️
+
+If you have a project-level .npmrc file and have added it in your version control then you could do the following:
+
+**Use environment variables**
+
+we could export the authentication string as an environment variable:
+
+e.g. for linux `$ export NPM_TOKEN=<base64-string>`
+
+We can then add the env variable in our .npmrc file:
+
+```
+//registry.wombatcorp.org/:_auth=${NPM_TOKEN}
+```
+
+The npm CLI will replace this value with the contents of the NPM_TOKEN environment variable
+    
+**Add scoped registry to user-level instead of project-level**
```

Comment:
with the change above then definitely :grin:

Reply:
I removed the whole paragraph about different npmrc files. Like I said above I wanted to emphasize that there are more possibilities how to not save the credentials in git and, to not leave the reader alone searching for solutions, at least, put 2 possibilities there. But the second block sounds a bit out of context if we remove the first headline. And if you didn't understand the listing then the usual reader won't understand either :smile: so let's go with one possibility what jeff does because this is also valuable information that npm replaces environment variables.